### PR TITLE
Feature 68

### PR DESCRIPTION
**Pull request by anna-riesch**
Introduces cli option to set a timeout for RPCs

### CODE REVIEW

File: src/libCli/Call.cpp [hunk 1]
```diff
@@ -146,7 +147,31 @@ namespace cli
         // Prepare the RPC call:
         std::multimap<grpc::string, grpc::string> clientMetadata;
         std::string methodStr = "/" + serviceName + "/" + methodName;
-        grpc::testing::CliCall call(channel, methodStr, clientMetadata);
+
+        // Get deadline for RPC from input or use custom
+        std::optional<std::chrono::time_point<std::chrono::system_clock>> timeout;
+        std::chrono::time_point<std::chrono::system_clock> defaultTimeout = std::chrono::system_clock::now() + std::chrono::milliseconds(100);
```
Comment (rainerschoe) [inline src/libCli/Call.cpp:153]: be consistent in naming: `chrono::time_point`s are `deadlines` and `chrono::durations` are `timeouts`

File: src/libCli/Call.cpp [hunk 2]
```diff
@@ -146,7 +147,31 @@ namespace cli
...
+                // Pass no timeout
+                timeout = std::nullopt;
+            }else{
+                timeout = defaultTimeout;
+            }
+        }
+
+        if(setTimeout){
+            std::string customTimeout = parseTree.findFirstChild("rpcTimeoutInMs");
+            int customTimeoutMs = std::stoi(customTimeout);
```
Comment (rainerschoe) [inline src/libCli/Call.cpp:169]: what happens ifstoi fails to parse? did you test this?

File: src/libCli/GrammarConstruction.cpp [hunk 3]
```diff
@@ -519,6 +519,10 @@ namespace cli
         optionsalt->addChild(jsonInput);
         optionsalt->addChild(f_grammarPool.createElement<FixedString>("--printParsedMessage", "PrintParsedMessage"));
         optionsalt->addChild(f_grammarPool.createElement<FixedString>("--noSimpleMapOutput", "NoSimpleMapOutput"));
+        GrammarElement *timeout = f_grammarPool.createElement<Concatenation>();
+        timeout->addChild(f_grammarPool.createElement<FixedString>("--rpcTimeoutMilliseconds=", "rpcTimeout"));
+        timeout->addChild(f_grammarPool.createElement<RegEx>("[0-9]+", "rpcTimeoutInMs"));
```
Comment (rainerschoe) [inline src/libCli/GrammarConstruction.cpp:524]: as states above. maybe add an alternation here to either have the regex or a fixed string `none`

File: tests/functionTests/rpcTimeoutFuntionTest.txt [hunk 4]
```diff
@@ -0,0 +1,79 @@
+# Copyright 2022 IBM Corporation
```
Comment (rainerschoe) [inline tests/functionTests/rpcTimeoutFuntionTest.txt:1]: typo in filename "FunctionTest"

File: tests/functionTests/rpcTimeoutFuntionTest.txt [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+@@CMD@@ --rpcTimeoutMilliseconds=500 localhost:50051 examples.StreamingRpcs bidirectionalStreamEchoNumberOrString :number=5:
+/.* Received message:
+| number = 5
+| str... = [NOT SET]
+| both.. = [NOT SET]
+RPC succeeded :D
+#END_TEST
+
+#START_TEST streaming RPC exeeding custom timeout
+@@CMD@@ --rpcTimeoutMilliseconds=500 localhost:50051 examples.StreamingRpcs bidirectionalStreamInfiniteRpc
```
Comment (rainerschoe) [inline tests/functionTests/rpcTimeoutFuntionTest.txt:77]: there is AFAIK a test server RPC which is outputting the current time in a specified frequency. You could use this to more acurately test streamign timeouts in addition to never ending RPC

File: tests/testServer/ServiceStreamingRpcs.cpp [hunk 6]
```diff
@@ -116,3 +116,36 @@ ::grpc::Status ServiceStreamingRpcs::bidirectionalStreamEchoNumberOrString(
     }
     return grpc::Status();
 }
+
+//::grpc::Status ServiceStreamingRpcs::longRunningStream(
```
Comment (rainerschoe) [inline tests/testServer/ServiceStreamingRpcs.cpp:120]: avoid dead code

File: tests/testServer/testServer.cpp [hunk 7]
```diff
@@ -190,6 +190,11 @@ int main(int argc, char **argv)
     ServiceStatusHandling statusHandling;
     builder.RegisterService(&statusHandling);
 
+
```
Comment (rainerschoe) [inline tests/testServer/testServer.cpp:193]: remove

**Review by rainerschoe — CHANGES_REQUESTED** [review]
Looks good :) some changes and additions and we are good to go :)

File: src/libCli/Call.cpp (hunk 2 above)
Comment (anna-riesch) [reply src/libCli/Call.cpp:169]: Do you mean e.g. if someone tries to provide a string as custom timeout? If so, the parse fails.

Comment (rainerschoe) [reply src/libCli/Call.cpp:169]: yes exactly, so I woule expect a functional test implemented testing this and a readable error to be printed

**Review by rainerschoe — APPROVED** [review]

**Pull request merged**